### PR TITLE
Fixes #3139 Attachment: a bug and a mod to allow multi-attachment selection

### LINKED ISSUE #3139: Attachment: a bug and a mod to allow multi-attachment selection
First, the bug (1.8.14).  The situation: one has uploaded an attachment for a new/edit post and one has selected another file to be uploaded but "Add Attachment to upload the second attachment has not been clicked yet.  If at that point, one clicks on "Remove Attachment", the selected attachment will be uploaded. 

Now an improvement.  I was interested in allowing multi-attachments adds.  I found a plugin that added that capability, but it also added a number of other features.  I found the plugin unusable because of various bugs, many of which were related to features other than supporting multifile upload.  I implemented the following fairly straight-forward changes to allow the multi-file uploads.  If anyone wants to try it, and provide feedback, at some point I may turn it into a plugin.

Uploaded files come into the *.php code via the variables "&#36;_FILES['attachment'][&#36;field]", where &#36;field has various values like 'name' and 'size' for various aspects of the uploaded file. For multi-file uploads, those variables become arrays instead of singular values.  Where those variables are used, the following code segment must be inserted.

Code:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&#36;fields = array ('name', 'type', 'tmp_name', 'error', 'size');
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if (is_array(&#36;_FILES['attachment']['name']))
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;{
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;// Already in multi-attachment array format
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&#36;attachments = &#36;_FILES;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;else
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;{
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;// Convert original-style non-array &#36;_FILES['attachment'][&#36;field] to array format in &#36;attachments
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&#36;attachments = array('attachment' =&gt; array());
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;foreach (&#36;fields as &#36;field)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&#36;attachments['attachment'][&#36;field] = array(&#36;_FILES['attachment'][&#36;field]);
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;foreach (&#36;attachments['attachment']['name'] as &#36;key =&gt; &#36;name)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;{
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;// Convert array &#36;attachments['attachment'][&#36;field][&#36;key] to non-array format in &#36;FILE
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;    &#36;FILE = array('attachment' =&gt; array());&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;    foreach (&#36;fields as &#36;field)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&#36;FILE['attachment'][&#36;field] = &#36;attachments['attachment'][&#36;field][&#36;key];
Step 1: The affected files are newthread.php, editpost.php and newreply.php, around line 225, give or take, in each.  In each file, insert the code before the first "if" statements that process the attachment (and add another closing curly brace for the added "foreach" statement). 

Step 2: In that "if" staement, chnage all references to "&#36;_FILES" to "&#36;FILE".

Example: in newthread.php, the new code segment would look like this:
[attachment=40094]

Step 4: Change the "input" tag in template "post_attachments_new" to have name "attachment[]" (braces added) and the keyword "multiple".
Code:&lt;input type="file" name="attachment[]" size="30" class="fileupload" multiple/&gt;
Older browsers may still be limited to single attachment uploads but it still should work for that.

Original thread: [Attachment: a bug and a mod to allow multi-attachment selection](https://community.mybb.com/thread-216548.html)

### PR DESCRIPTION
**Pull request by Sama34**
Fixes #3139 Attachment: a bug and a mod to allow multi-attachment selection

Please review.

### CODE REVIEW

**Comment by SvePu** [comment]
@Sama34 Thx for this nice fix, but you'll get an MySql syntax error on editpost because $attachwhere is empty

**Comment by SvePu** [comment]
@Sama34 - also I get an empty($_POST) error when summary of all selected files sizes is greater than php post_max_size. May be we need a size check before start uploading

Could someone confirm this?

File: install/resources/mybb_theme.xml [hunk 1]
```diff
@@ -9459,7 +9452,7 @@ if(use_xmlhttprequest == "1")
 </tr>]]></template>
 		<template name="post_attachments_new" version="1807"><![CDATA[<tr>
```
Comment (euantorano) [inline install/resources/mybb_theme.xml]: This template needs the version bumping to `1816`.

**Comment by SvePu** [comment]
@Sama34 
I fixed the issue by replacing in editpost.php:
```
$ret = add_attachments($pid, $forumpermissions, $attachwhere, "editpost");
```
with:
```
$ret = add_attachments($pid, $forumpermissions, "pid='{$pid}'", "editpost");
```
@euantorano 
I use some litte JavaScript for this size check - if JavaScript is disabled in browser the multiupload option will be changed into a single upload.

**Comment by euantorano** [comment]
@SvePu That sounds like a good option to me.

**Comment by SvePu** [comment]
Maybe someone with more knowledge in JavaScript could integrate something like this into jscripts/post.js
The vars for max_upload_size and post_max_size we could get over:
```
ini_get('upload_max_filesize');
```
and
```
ini_get('post_max_size');
```

**Comment by Eldenroot** [comment]
@Sama34 - can you please incorporate fixes above? I tested this PR and seems to be fine except these two things. Thank you'

**Comment by Ben-MyBB** [comment]
@Sama34 ping

**Comment by effone** [comment]
Can you please look at the PR @Sama34 ? Seems to be the last PR for 1.8.16 so far ...

**Comment by Eldenroot** [comment]
Or is there anybody who can merge this and add the requested chnages in next PR?

**Comment by effone** [comment]
May be @SvePu can create an extended part of this and then we can merge both back to back

**Comment by SvePu** [comment]
@effone 
I use this small JS inline script to check upload size and file count before fire up the upload. Maybe someone could integrate something like this into post.js to call it on click at submit button.

``` javascript
<script type="text/javascript">	
	$(function(){
		document.getElementById("file").setAttribute("multiple","multiple");
		$("input[type='submit']").click(function(){
			var fileUpload = $("input[type='file']");
			if (parseInt(fileUpload.get(0).files.length)>{$php_max_file_uploads}){
				alert("Du kannst maximal {$php_max_file_uploads} Dateien gleichzeitig hochladen - Führe wenn nötig mehrere Durchgänge aus.");
				document.getElementById("file").value="";
				return false;
			}
			var totalSize = 0;
			fileUpload.each(function() {
				for (var i = 0; i < this.files.length; i++) {
					totalSize += this.files[i].size;
				}
			});
			var valid = totalSize <= {$php_max_upload_size};
			if (!valid) {
				alert('Die hochzuladenen Dateien haben eine Gesamtgröße über {$php_max_upload_size_mb} - bitte führe wenn nötig mehrere Durchgänge aus.');
				document.getElementById("file").value="";
				return false;
			}
		});
	});
</script>
```
Sorry....the alert info is in German language ;)

**Comment by Eldenroot** [comment]
Is there anybody who can make a new PR with JS script above? I would like to but my JS knowledge is almost zero :(

**Comment by euantorano** [comment]
@Eldenroot This PR won't be making it into 1.8.16 unfortunately, it'll have to wait until .17.

**Comment by Eldenroot** [comment]
I use this "new feature" on my board (without JS part), I applied changed from SvePu, working fine. Is there anybody who can take care of JS part and add it inside post.js? @Sama34

**Comment by euantorano** [comment]
Superseded by #3373.

**Pull request closed without merging**